### PR TITLE
add support for instance checks on dataset wrappers

### PR DESCRIPTION
**Pull request by pmeier**
Per title following an offline discussion with @NicolasHug. 

cc @vfdev-5 @bjuncek

### CODE REVIEW

File: torchvision/datasets/vision.py [hunk 1]
```diff
@@ -7,7 +9,17 @@
 from ..utils import _log_api_usage_once
 
 
-class VisionDataset(data.Dataset):
+class Hack(ABCMeta):
```
Comment (pmeier) [inline torchvision/datasets/vision.py:12]: We just need to subclass from `ABCMeta` here, since some of our datasets inherit from `ABC`:

https://github.com/pytorch/vision/blob/707457050620e1f70ab1b187dad81cc36a7f9180/torchvision/datasets/_optical_flow.py#L30

Without them, we could also inherit from `type` directly.

Comment (NicolasHug) [reply torchvision/datasets/vision.py:12]: Could we apply something like that to the `VisionDatasetDatapointWrapper` class instead of altering all the existing classes?

Comment (pmeier) [reply torchvision/datasets/vision.py:12]: I don't think so, no. `__instancecheck__` is only triggered for the second argument of the `instance` call. Meaning, we have no hooks for `isinstance(wrapped_mnist, datasets.MNIST)` unless we change something about `datasets.MNIST` or its base class `datasets.VisionDataset`.

The only other way to achieve that is to let our wrapper properly subclass. Let me give it a shot.

File: torchvision/prototype/datapoints/_dataset_wrapper.py [hunk 2]
```diff
@@ -17,7 +16,7 @@
 
 # TODO: naming!
 def wrap_dataset_for_transforms_v2(dataset):
-    return VisionDatasetDatapointWrapper(dataset)
+    return type(f"Wrapped{type(dataset).__name__}", (VisionDatasetDatapointWrapper, type(dataset)), {})(dataset)
```
Comment (pmeier) [inline torchvision/prototype/datapoints/_dataset_wrapper.py:19]: This is the minimal change to get the `isinstance` check working without touching the original dataset code at all. I feel this is a little more hacky than the previous meta class approach, but this may be justified for this (hopefully) temporary wrapper.

We dynamically create a class here and instantiate it. For example, if the input is a `CocoDetection`, this is equivalent to

```python
class WrappedCocoDetection(VisionDatasetDatapointWrapper, CocoDetection):
    pass
```

Due to the double inheritance, `isinstance(dataset, CocoDetection)` now passes. However, we still use composition internally, since we hit the constructor of `VisionDatasetDatapointWrapper` and never call `super().__init__`.

**Review by Dmatt18 — APPROVED** [review]

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip
This looks a bit insane but [for now] I'm not against it. It would help users integrate their code more smoothly, as we experienced in https://github.com/pytorch/vision/pull/7732

What else do we need to change to make it happen?

File: torchvision/datapoints/_dataset_wrapper.py [hunk 3]
```diff
@@ -16,7 +15,7 @@
 
 # TODO: naming!
 def wrap_dataset_for_transforms_v2(dataset):
-    return VisionDatasetDatapointWrapper(dataset)
+    return type(f"Wrapped{type(dataset).__name__}", (VisionDatasetDatapointWrapper, type(dataset)), {})(dataset)
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/datapoints/_dataset_wrapper.py:18]: > OK so we're creating a class on the fly named `WrappedImageNet` that inherits from both `VisionDatasetDatapointWrapper` _and_ `ImageNet`, and we instanciate it with `ImageNet()` which is in fact passed to the constructor of `VisionDatasetDatapointWrapper`?

Correct. I think that is what you meant, but the `ImageNet()` instance is what we get from the user. Meaning, we never actually create a dataset object inside the wrapper.

> So `isinstance(ds, ImageNet)` and `isinstance(ds, VisionDatasetDatapointWrapper)` and `isinstance(ds, WrappedImageNet)` are all true (except that `WrappedImageNet` isn't in scope anywhere)?

Correct.

> (except that `WrappedImageNet` isn't in scope anywhere)

Well that is what users get when they do `type(wrapped_dataset)`, but I'm ok with that since it is not really different from before where this returned `VisionDatasetDatapointWrapper`. This also is never exposed anywhere. Meaning, if users only use `isinstace` we are good. If they go near `type` it will be complicated.

> I assume that this doesn't work:
> 
> ```python
> type(f"Wrapped{type(dataset).__name__}", (type(dataset),), {})(VisionDatasetDatapointWrapper(dataset))
> ```
> 
> because the `ImageNet` constructor doesn't expect a `VisionDatasetDatapointWrapper()` instance?

Correct. If we inherit from `ImageNet` and don't define a constructor, the one from the superclass is called.

**Comment by pmeier** [comment]
> What else do we need to change to make it happen?

Nothing, this should work as is. One thing that I could try is to eliminate the `VisionDatasetDatapointWrapper` class.

File: torchvision/datapoints/_dataset_wrapper.py [hunk 4]
```diff
@@ -98,7 +97,16 @@ def wrap_dataset_for_transforms_v2(dataset, target_keys=None):
...
 
-    return VisionDatasetDatapointWrapper(dataset, target_keys)
+    return type(
+        f"Wrapped{type(dataset).__name__}",
+        (type(dataset),),
+        {
+            "__init__": wrapped_init,
+            "__getattr__": wrapped_getattr,
+            "__getitem__": wrapped_getitem,
+            "__len__": wrapped_len,
```
Comment (NicolasHug) [inline torchvision/datapoints/_dataset_wrapper.py:107]: Why do we need this? Shouldn't the MRO do its job without us having to override anything here?

Comment (NicolasHug) [reply torchvision/datapoints/_dataset_wrapper.py:107]: Oh, if it's because we have now removed the `VisionDatasetDatapointWrapper`, let's just not remove it please. At least not in this PR. (we can keep it with a leading underscore if we want to).

File: test/datasets_utils.py [hunk 5]
```diff
@@ -584,8 +584,10 @@ def test_transforms_v2_wrapper(self, config):
                         continue
 
                     wrapped_dataset = wrap_dataset_for_transforms_v2(dataset, target_keys=target_keys)
-                    wrapped_sample = wrapped_dataset[0]
+                    assert isinstance(wrapped_dataset, self.DATASET_CLASS)
+                    assert len(wrapped_dataset) == info["num_examples"]
```
Comment (NicolasHug) [inline test/datasets_utils.py:588]: is that related or is it a drive-by?

File: torchvision/datapoints/_dataset_wrapper.py [hunk 6]
```diff
@@ -98,7 +97,16 @@ def wrap_dataset_for_transforms_v2(dataset, target_keys=None):
             f"but got {target_keys}"
         )
 
-    return VisionDatasetDatapointWrapper(dataset, target_keys)
+    # Imagine we have isinstance(dataset, datasets.ImageNet). This will create a type with the name "WrappedImageNet" at
```
Comment (NicolasHug) [inline torchvision/datapoints/_dataset_wrapper.py:100]: I know it's the same but since the variable name is `wrapped_dataset_cls` this will be slightly more obvious.

```suggestion
    # Imagine we have isinstance(dataset, datasets.ImageNet). This will create a new class with the name "WrappedImageNet" at
```

**Review by NicolasHug — APPROVED** [review]
Thanks Philip.

Do we have a lot of datasets that store extra attributes

File: test/datasets_utils.py (hunk 5 above)
Comment (pmeier) [reply test/datasets_utils.py:588]: L587 is needed to enforce the `isinstance` check works properly. L588 is a driveby.
